exp/addons/controllers/clusterresourcesetbinding_controller.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	addonsv1 "sigs.k8s.io/cluster-api/exp/addons/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util/predicates"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// +kubebuilder:rbac:groups=addons.cluster.x-k8s.io,resources=*,verbs=get;list;watch;create;update;patch;delete
+
+// ClusterResourceSetBindingReconciler reconciles a ClusterResourceSetBinding object
+type ClusterResourceSetBindingReconciler struct {
+	Client client.Client
+	Log    logr.Logger
+
+	scheme *runtime.Scheme
+}
+
+func (r *ClusterResourceSetBindingReconciler) SetupWithManager(mgr ctrl.Manager, options controller.Options) error {
+	_, err := ctrl.NewControllerManagedBy(mgr).
+		For(&addonsv1.ClusterResourceSetBinding{}).
+		Watches(
+			&source.Kind{Type: &clusterv1.Cluster{}},
+			&handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(r.clusterToClusterResourceSetBinding)},
+		).
```

Comment:
This cluster watch should probably be configured separately similarly to what we are doing in the Machine controller to properly honor paused clusters (at least until we can use a newer version of controller-runtime that supports configuring individual predicates for a watch):
https://github.com/kubernetes-sigs/cluster-api/blob/60a6daf4b9ecb74a9111b3269e7016988a8afc00/controllers/machine_controller.go#L99-L109

As a side note, I think this applies to the ClusterResourceSet controller as well.

Reply:
I tried this before, but somehow it does not trigger cluster handler. I may be missing something. clusterToClusterResourceSetBinding() never gets called. @detiber

Reply:
Ah, that is correct, the default predicates we define wont work for this use case.